### PR TITLE
#18059 #15687 Enable https hostname verification 

### LINKED ISSUE #18059: Need access to `createSslEngine(hostname, port)` to fix #15687
The SSL engine currently uses only the no-arg method. However, the other one is needed for Java 7's `setEndpointIdentificationAlgorithm` to be used to fix #15687.

We could maybe abuse `Role` for this purpose by providing a `Role.ClientOf(hostname, port)`?

### PR DESCRIPTION
**Pull request by jrudolph**
This is a (not necessarily complete) suggestion how to fix the issue. Adding `hostInfo` has the additional benefit that sessions may then be reused speeding up subsequent connection attempts. It has the slight disadvantage that you cannot reuse a SslTls BidiFlow for connections to different hosts.

Fixes #18059, #15687.

/cc @sirthias @rkuhn

### CODE REVIEW

**Comment by jrudolph** [comment]
Error lines in https://jenkins.akka.io/job/pr-validator-per-commit-jenkins/3274/:

```
 10695 Failed: Total 457, Failed 1, Errors 0, Passed 456, Pending 20
 10696 Failed tests:
 10697 	akka.http.impl.engine.client.TlsEndpointVerificationSpec
 14060 (akka-http-core-experimental/test:test) sbt.TestsFailedException: Tests unsuccessful
 14061 Total time: 838 s, completed Jul 23, 2015 3:53:24 PM
```

File: akka-http-core/src/main/resources/keys/README [hunk 1]
```diff
@@ -0,0 +1,39 @@
...
+# Create certificate chain:
+
+cat server.crt rootCA.crt > chain.pem
+
+# Convert certificate and key to pkcs12 (you need to provide a password manually, `ExampleHttpContexts`
+# expects the password to be "abcdef"):
+
+openssl pkcs12 -export -name servercrt -in chain.pem -inkey server.key -out server.p12
+
+
```
Comment (ktoso) [inline akka-http-core/src/main/resources/keys/README]: thanks for the instructions!

File: akka-http-core/src/main/scala/akka/http/impl/util/Java6Compat.scala [hunk 2]
```diff
@@ -0,0 +1,53 @@
...
+
+import java.lang.reflect.{ InvocationTargetException, Method }
+import javax.net.ssl.SSLParameters
+
+import scala.util.control.NonFatal
+
+/**
+ * Enables accessing SslParameters even if compiled against Java 6.
+ *
+ * Internal API
```
Comment (ktoso) [inline akka-http-core/src/main/scala/akka/http/impl/util/Java6Compat.scala]: nitpick: in first line and `INTERNAL API`

File: akka-http-core/src/main/scala/akka/http/impl/util/Java6Compat.scala [hunk 3]
```diff
@@ -0,0 +1,53 @@
...
+        } else unsupported
+
+      // probe so that we can be sure a reflective problem only turns up once
+      // here during construction
+      candidate(new SSLParameters(), "https")
+      candidate
+    } catch {
+      case NonFatal(_) ⇒ unsupported
+    }
+  }
```
Comment (ktoso) [inline akka-http-core/src/main/scala/akka/http/impl/util/Java6Compat.scala]: looks good

File: akka-http-core/src/main/scala/akka/http/impl/util/Java6Compat.scala (hunk 2 above)
Comment (jrudolph) [inline akka-http-core/src/main/scala/akka/http/impl/util/Java6Compat.scala]: Thanks fixed.

File: akka-docs-dev/rst/java/http/client-side/https-support.rst [hunk 4]
```diff
@@ -37,5 +37,20 @@ Afterwards you simply use ``outgoingConnectionTls``, ``newHostConnectionPoolTls`
...
+-------------------------------
+
+Hostname verification proves that the Akka HTTP client is actually communicating with the server it intended to
+communicate with. Without this check a man-in-the-middle attack is possible. In the attack scenario, an alternative
+certificate would be presented which was issued for another host name. Checking the host name in the certificate
+against the host name the connection was opened against is therefore vital.
+
+The default ``HttpsContext`` enables hostname verification. Akka HTTP relies on a Java 7 feature to implement
+the verification. To prevent an unintended security downgrade, accessing the default ``HttpsContext`` on Java 6
+will fail with an exception. Specifying a custom ``HttpsContext`` or customizing the default one is possible also Java 6.
```
Comment (2beaucoup) [inline akka-docs-dev/rst/java/http/client-side/https-support.rst]: "... is possible also Java 6" (2x)

File: akka-http-core/src/main/scala/akka/http/scaladsl/Http.scala [hunk 5]
```diff
@@ -390,13 +390,25 @@ class HttpExt(config: Config)(implicit system: ActorSystem) extends akka.actor.E
...
         case ctx ⇒ ctx
       }
     }
 
+  private def createDefaultClientHttpsContext: HttpsContext = {
+    val defaultCtx = SSLContext.getDefault
+
+    val params = new SSLParameters
+    if (!Java6Compat.setEndpointIdentificationAlgorithm(params, "https"))
+      throw new ReadTheDocumentationException(
```
Comment (2beaucoup) [inline akka-http-core/src/main/scala/akka/http/scaladsl/Http.scala]: Why not use `RuntimeException` directly?

File: akka-http-core/src/main/scala/akka/http/impl/util/Java6Compat.scala [hunk 6]
```diff
@@ -0,0 +1,53 @@
...
+import javax.net.ssl.SSLParameters
+
+import scala.util.control.NonFatal
+
+/**
+ * INTERNAL API
+ *
+ * Enables accessing SslParameters even if compiled against Java 6.
+ */
+private[http] object Java6Compat {
```
Comment (2beaucoup) [inline akka-http-core/src/main/scala/akka/http/impl/util/Java6Compat.scala]: :+1: There are definetly some more Candidates to be moved here. Would be nice to have a central place for them so we can get rid of the legacy stuff easier.

Comment (2beaucoup) [inline akka-http-core/src/main/scala/akka/http/impl/util/Java6Compat.scala]: Mmh, we already have `EnhancedInetSocketAddress.getHostStringJava6Compatible`.

Would `EnhancedSSLParameters.setEndpointIdentificationAlgorithmJava8Only` be an option here?

Comment (jrudolph) [inline akka-http-core/src/main/scala/akka/http/impl/util/Java6Compat.scala]: I thought about an `EnhancedSSLParameters`, too, but it seems a little overkill for a one-time use.

`getHostStringJava6Compatible` is used several times, but still I would rather move `getHostStringJava6Compatible` here (for the reasons you mentioned above) than the other way round.

Comment (2beaucoup) [inline akka-http-core/src/main/scala/akka/http/impl/util/Java6Compat.scala]: Having thought once more about it (and after searching for mentions of "getHostString") I think that we should even move `getHostStringJava6Compatible` to `akka.io` or `akka.util` as there are a lot of `TODO`s all over the place.

File: akka-http-core/src/main/scala/akka/http/scaladsl/Http.scala (hunk 5 above)
Comment (ktoso) [inline akka-http-core/src/main/scala/akka/http/scaladsl/Http.scala]: I don't mind the `ReadTheDocumentationException`, sounds good (hey, it's not RTFM ;-))

File: akka-docs-dev/rst/java/http/client-side/https-support.rst (hunk 4 above)
Comment (ktoso) [inline akka-docs-dev/rst/java/http/client-side/https-support.rst]: ?

I thought wording should be "is also possible on Java 6"

**Comment by ktoso** [comment]
LGTM, not sure if you'd like to address @2beaucoup's comment right away here?

File: akka-http-core/src/main/scala/akka/http/impl/util/Java6Compat.scala (hunk 6 above)
Comment (2beaucoup) [inline akka-http-core/src/main/scala/akka/http/impl/util/Java6Compat.scala]: #18098

**Comment by 2beaucoup** [comment]
I created a seperate issue re. `EnhancedInetSocketAddress.getHostStringJava6Compatible`.

**Comment by jrudolph** [comment]
Thanks, @2beaucoup for creating the ticket.

@ktoso I think we should keep those issues separate to be able to merge this ASAP.

I'll fix the remaining docs issues.

**Comment by ktoso** [comment]
@jrudolph, yeap agreed - will merge as soon as nitpicks fixed then :)

File: akka-docs-dev/rst/java/http/client-side/https-support.rst (hunk 4 above)
Comment (jrudolph) [inline akka-docs-dev/rst/java/http/client-side/https-support.rst]: Fixed.

**Comment by ktoso** [comment]
Failure was unrelated (timing issue) https://github.com/akka/akka/issues/18110

**Pull request merged**